File: widgets/widget_binder/widget.py
"""
WidgetBinder — a tiny anywidget that wires a source widget's trait change to a
target widget's trait, entirely in JS at render time. Used to demonstrate
cross-widget interop in MyST static export, e.g. "counter value drives lonboard
map zoom" with no kernel.

The binding is identified by:
  source_widget_id     — looked up in window.__myst_widgets (matches widget_id,
                         _anywidget_id, or model_id)
  source_field         — trait name to subscribe to on the source
  target_widget_id     — same lookup as source
  target_field         — dotted path on target ("view_state.zoom" merges into
                         the existing view_state object)
  multiplier, offset   — linear transform applied to the source value
                         (target = source * multiplier + offset)
"""
import anywidget
import traitlets
import pathlib


class WidgetBinder(anywidget.AnyWidget):
    _esm = pathlib.Path(__file__).parent / "widget.js"

    source_widget_id = traitlets.Unicode("").tag(sync=True)
    source_field = traitlets.Unicode("value").tag(sync=True)
    target_widget_id = traitlets.Unicode("").tag(sync=True)
    target_field = traitlets.Unicode("").tag(sync=True)
    multiplier = traitlets.Float(1.0).tag(sync=True)
    offset = traitlets.Float(0.0).tag(sync=True)
    label = traitlets.Unicode("").tag(sync=True)

File: widgets/widget_binder/widget.js
// WidgetBinder: subscribe to a source widget trait change and write to a
// target widget's trait through the AFM-style host.waitForModel API.
//
// In a JupyterLab kernel context this widget does nothing useful. Python
// observers are the right tool there. This is for static export, where there's
// no kernel and we need a JS-only binding.

function resolveModel(host, id, timeout = 5000) {
    if (!host || typeof host.waitForModel !== "function") {
        return Promise.reject(new Error("[binder] host.waitForModel is unavailable"));
    }
    return host.waitForModel(id, { timeout });
}

// Set a value at a dotted path on a target model. For leaf paths we simply
// model.set(key, value). For nested paths (e.g. "view_state.zoom"), we read
// the top-level object, merge the leaf, and set it back so listeners see the
// change as a single update.
function setByPath(model, path, value) {
    const parts = path.split(".");
    if (parts.length === 1) {
        model.set(parts[0], value);
        return;
    }
    const topKey = parts[0];
    const existing = model.get(topKey);
    const next = (existing && typeof existing === "object") ? { ...existing } : {};
    let cursor = next;
    for (let i = 1; i < parts.length - 1; i++) {
        const k = parts[i];
        cursor[k] = (cursor[k] && typeof cursor[k] === "object") ? { ...cursor[k] } : {};
        cursor = cursor[k];
    }
    cursor[parts[parts.length - 1]] = value;
    model.set(topKey, next);
}

function render({ model, el, host }) {
    const sourceId = model.get("source_widget_id");
    const sourceField = model.get("source_field") || "value";
    const targetId = model.get("target_widget_id");
    const targetField = model.get("target_field");
    const multiplier = model.get("multiplier");
    const offset = model.get("offset");
    const label = model.get("label") || `${sourceId}.${sourceField} → ${targetId}.${targetField}`;

    el.style.fontFamily = "ui-monospace, SFMono-Regular, monospace";
    el.style.fontSize = "12px";
    el.style.padding = "8px 12px";
    el.style.borderRadius = "6px";
    el.style.background = "#f6f8fa";
    el.style.color = "#24292e";
    el.style.border = "1px solid #e1e4e8";
    el.style.maxWidth = "fit-content";

    const status = document.createElement("div");
    el.appendChild(status);
    status.textContent = `🔗 binder: waiting…  (${label})`;

    Promise.all([
        resolveModel(host, sourceId, 5000),
        resolveModel(host, targetId, 5000),
    ]).then(([source, target]) => {
        const apply = () => {
            const raw = source.get(sourceField);
            const next = (typeof raw === "number") ? raw * multiplier + offset : raw;
            setByPath(target, targetField, next);
        };
        source.on(`change:${sourceField}`, apply);
        apply(); // also fire once at startup so the target reflects current source state
        status.textContent = `✅ ${label}`;
    }).catch(err => {
        status.textContent = `❌ binder failed: ${err.message}`;
        console.error("[binder] failed", err, { sourceId, targetId });
    });
}

export default { render };
